pkg/plugin/plugin.go

```diff
@@ -116,142 +136,119 @@ func getPluginConfig(rollout *v1alpha1.Rollout) (*GlooEdgeTrafficRouting, error)
+						routeTables[i].Destinations[j].Stable,
+					},
+				},
+			}
+		}
+	}
+}
 
-	if !strings.EqualFold(pluginConfig.RouteTableSelector.Name, "") {
-		r.LogCtx.Debugf("getRouteTables using ns:name ref %s:%s to get single table", pluginConfig.RouteTableSelector.Name, pluginConfig.RouteTableSelector.Namespace)
-		result, err := r.Client.RouteTables().GetRouteTable(ctx, pluginConfig.RouteTableSelector.Name, pluginConfig.RouteTableSelector.Namespace)
-		if err != nil {
-			return nil, err
+func (r *RpcPlugin) maybeCreateCanaryDestinations(
+	routeTables []routeTableWithDestinations, canaryName string) {
+
+	for i := range routeTables {
+		for j := range routeTables[i].Destinations {
+			if routeTables[i].Destinations[j].Canary != nil {
+				continue
+			}
+			routeTables[i].Destinations[j].Canary =
+				r.newCanaryDestination(routeTables[i].Destinations[j].Stable, canaryName)
+			routeTables[i].Destinations[j].DestinationsParent.GetMulti().Destinations =
+				append(routeTables[i].Destinations[j].DestinationsParent.GetMulti().GetDestinations(), routeTables[i].Destinations[j].Canary)
 		}
+	}
+}
 
-		r.LogCtx.Debugf("getRouteTables using ns:name ref %s:%s found 1 table", pluginConfig.RouteTableSelector.Name, pluginConfig.RouteTableSelector.Namespace)
-		rts = append(rts, result)
-	} else {
-		opts := &k8sclient.ListOptions{}
+func (r *RpcPlugin) newCanaryDestination(stableDst *v1.WeightedDestination, canaryName string) *v1.WeightedDestination {
+	ret := stableDst.Clone().(*v1.WeightedDestination)
+	ret.GetDestination().GetUpstream().Name = canaryName
+	ret.Weight = &wrapperspb.UInt32Value{Value: uint32(0)}
+	return ret
+}
 
-		if pluginConfig.RouteTableSelector.Labels != nil {
-			opts.LabelSelector = labels.SelectorFromSet(pluginConfig.RouteTableSelector.Labels)
+func (r *RpcPlugin) getDestinationsInRoutes(
+	routes []*gwv1.Route,
+	rollout *v1alpha1.Rollout,
+	pluginConfig *GlooEdgeTrafficRouting) (ret []destinationPair) {
+
+	for _, route := range routes {
+		if len(pluginConfig.Routes) > 0 && !slices.Contains(pluginConfig.Routes, route.GetName()) {
+			continue
 		}
-		if !strings.EqualFold(pluginConfig.RouteTableSelector.Namespace, "") {
-			opts.Namespace = pluginConfig.RouteTableSelector.Namespace
+
+		if route.GetRouteAction() == nil ||
+			(route.GetRouteAction().GetMulti() == nil && route.GetRouteAction().GetSingle() == nil) {
+			continue
 		}
 
-		r.LogCtx.Debugf("getRouteTables listing tables with opts %+v", opts)
-		var err error
+		if route.GetRouteAction().GetSingle() != nil {
+			ret = append(ret, r.getDestinationInSingle(route, rollout)...)
+			continue
+		}
 
-		rts, err = r.Client.RouteTables().ListRouteTables(ctx, opts)
-		if err != nil {
-			return nil, err
+		if route.GetRouteAction().GetMulti().GetDestinations() == nil {
+			continue
 		}
-		r.LogCtx.Debugf("getRouteTables listing tables with opts %+v; found %d routeTables", opts, len(rts))
+		ret = append(ret, r.getDestinationsInMulti(route, rollout)...)
```

Comment:
nit: I think some of this logic can be simplified. It seems like this logic attempts to do the following:

For each route:
- if a list of available routes is provided, the route name must be in the list
- if the route singleAction is defined, append the WeightedDestinations from that route
- If the route MultiActions is defined, append the WeightedDestinations
- continue

I think what you have is good as is, it just took some extra for me to undersatnd the logic and I figured, the following might be easier to reason through:

```go
  if route.GetRouteAction().GetSingle() != nil {
    ret = append(ret, r.getDestinationInSingle(route, rollout)...)
    continue
  }

  if route.GetRouteAction().GetMulti() != nil {
    ret = append(ret, r.getDestinationsInMulti(route, rollout)...)
    continue
  }
```

Reply:
Haha, yeah, I see what you mean, changed.